### PR TITLE
Style: apply ruff-isort rule, sort imports

### PR DESCRIPTION
**Pull request by connortann**
Making incremental progress towards a standard code style. Enables ruff rule "I" (as enforced via pre-commit); sort imports.

Following discussion: https://github.com/dsgibbons/shap/discussions/80#discussioncomment-6013063 : now that the SHAP Community Fork has been merged, I feel we do not need to worry so much about merge conflicts between forks.

### CODE REVIEW

**Comment by connortann** [comment]
As part of this PR I noticed an issue relating to scipy imports, so opened a new issue #2975.

This PR makes some progress towards fixing the issue, but I think a full fix is best kept for a separate PR that can be discussed separately.

File: shap/explainers/_explainer.py [hunk 1]
```diff
@@ -1,19 +1,17 @@
 import copy
 import time
+
 import numpy as np
 import scipy as sp
```
Comment (thatlittleboy) [inline shap/explainers/_explainer.py]: How about this scipy import (#2975)? It should be explicit as well right?

File: shap/explainers/_kernel.py [hunk 2]
```diff
@@ -1,21 +1,33 @@
...
+import gc
 import itertools
+import logging
 import warnings
-import gc
-from packaging import version
+
+import numpy as np
+import pandas as pd
+import scipy as sp
```
Comment (thatlittleboy) [inline shap/explainers/_kernel.py]: Should this be explicit as well ( #2975)?

File: shap/explainers/_linear.py [hunk 3]
```diff
@@ -1,12 +1,17 @@
+import warnings
+
 import numpy as np
 import scipy as sp
```
Comment (thatlittleboy) [inline shap/explainers/_linear.py]: ditto

File: shap/explainers/other/_maple.py [hunk 4]
```diff
@@ -139,10 +141,11 @@ def attributions(self, X, multiply_by_input=False):
 # -  Assumes any required data normalization has already been done
 # -  Can pass Y (desired response) instead of MR (model fit to Y) to make fitting MAPLE to datasets easy
 
-from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
+import numpy as np
+from sklearn.ensemble import GradientBoostingRegressor, RandomForestRegressor
 from sklearn.linear_model import Ridge
 from sklearn.metrics import mean_squared_error
-import numpy as np
+
```
Comment (thatlittleboy) [inline shap/explainers/other/_maple.py:148]: We shouldnt be touching the vendored code, as mentioned in #2968. Maybe lets merge in that PR first

File: shap/models/_teacher_forcing.py [hunk 5]
```diff
@@ -1,10 +1,11 @@
 import numpy as np
 import scipy as sp
```
Comment (thatlittleboy) [inline shap/models/_teacher_forcing.py]: Ditto about scipy import

File: shap/models/_teacher_forcing.py [hunk 6]
```diff
@@ -254,7 +255,7 @@ def model_inference(self, inputs, output_ids):
             Returns output logits from the model.
         """
         if self.similarity_model_type == "pt":
-            import torch # pylint: disable=import-outside-toplevel
+            import torch  # pylint: disable=import-outside-toplevel
```
Comment (thatlittleboy) [inline shap/models/_teacher_forcing.py:258]: Could we change this to the ruff rule equivalent?

File: shap/models/_teacher_forcing.py [hunk 7]
```diff
@@ -276,7 +277,7 @@ def model_inference(self, inputs, output_ids):
                     outputs = self.similarity_model(**inputs, return_dict=True)
                 logits = outputs.logits.detach().cpu().numpy().astype('float64')
         elif self.similarity_model_type == "tf":
-            import tensorflow as tf # pylint: disable=import-outside-toplevel
+            import tensorflow as tf  # pylint: disable=import-outside-toplevel
```
Comment (thatlittleboy) [inline shap/models/_teacher_forcing.py:280]: same here

File: shap/models/_text_generation.py [hunk 8]
```diff
@@ -150,7 +152,7 @@ def model_generate(self, X):
             # device = torch.device('cuda' if torch.cuda.is_available() else 'cpu') if self.device is None else self.device
             # self.inner_model = self.inner_model.to(device)
             # self.inner_model.eval()
-            import torch # pylint: disable=import-outside-toplevel
+            import torch  # pylint: disable=import-outside-toplevel
```
Comment (thatlittleboy) [inline shap/models/_text_generation.py:155]: ditto

File: shap/models/_text_generation.py [hunk 9]
```diff
@@ -168,7 +170,7 @@ def model_generate(self, X):
                 outputs = self.inner_model.generate(inputs, **text_generation_params).numpy()
             else:
                 try:
-                    import tensorflow as tf # pylint: disable=import-outside-toplevel
+                    import tensorflow as tf  # pylint: disable=import-outside-toplevel
```
Comment (thatlittleboy) [inline shap/models/_text_generation.py:173]: ditto

File: shap/models/_topk_lm.py [hunk 10]
```diff
@@ -1,9 +1,11 @@
 import numpy as np
 import scipy as sp
```
Comment (thatlittleboy) [inline shap/models/_topk_lm.py]: same comment about scipy import

File: shap/models/_topk_lm.py [hunk 11]
```diff
@@ -52,7 +54,7 @@ def __init__(self, model, tokenizer, k=10, generate_topk_token_ids=None, batch_s
         self.model_type = None
         if safe_isinstance(self.inner_model, "transformers.PreTrainedModel"):
             self.model_type = "pt"
-            import torch # pylint: disable=import-outside-toplevel
+            import torch  # pylint: disable=import-outside-toplevel
```
Comment (thatlittleboy) [inline shap/models/_topk_lm.py:57]: ditto. and a couple more in this file

File: shap/plots/_force.py [hunk 12]
```diff
@@ -2,32 +2,31 @@
...
 import json
+import os
 import random
+import re
+import string
+import warnings
+from collections.abc import Sequence
+
+import numpy as np
+import scipy as sp
```
Comment (thatlittleboy) [inline shap/plots/_force.py]: ditto about scipy import

File: shap/plots/_monitoring.py [hunk 13]
```diff
@@ -1,13 +1,15 @@
+import warnings
+
 import numpy as np
 import scipy
```
Comment (thatlittleboy) [inline shap/plots/_monitoring.py]: ditto here about scipy

File: shap/plots/colors/_colorconv.py [hunk 14]
```diff
@@ -121,9 +121,11 @@
 
 
 
+from warnings import warn
+
 import numpy as np
 from scipy import linalg
-from warnings import warn
+
```
Comment (thatlittleboy) [inline shap/plots/colors/_colorconv.py]: we shouldnt touch vendored code

File: shap/explainers/_explainer.py (hunk 1 above)
Comment (thatlittleboy) [reply shap/explainers/_explainer.py]: oops didnt see your comment. we can leave this to a separate PR if you wish.

File: shap/models/_teacher_forcing.py (hunk 6 above)
Comment (connortann) [reply shap/models/_teacher_forcing.py:258]: At present I don't think we have an equivalent Ruff rule, so I think we could leave as-is for now

File: shap/plots/colors/_colorconv.py (hunk 14 above)
Comment (connortann) [reply shap/plots/colors/_colorconv.py]: Good point, fixed

**Comment by connortann** [comment]
Thanks for the review comments. I've updated pushed some further commits to:

- Merge in master, resolving conflicts
- Update all the scipy submodude imports to fix #2975 , and
- Revert changes to vendored code

File: shap/models/_teacher_forcing.py (hunk 6 above)
Comment (connortann) [reply shap/models/_teacher_forcing.py:258]: Yep - From the Ruff list of codes: https://beta.ruff.rs/docs/rules/#pylint-pl , there isn't a code that corresponds to pylint `import-outside-toplevel / C0415` .

**Review by thatlittleboy — APPROVED** [review]
kitten-low-five-shipit

**Pull request merged**